File: solutions/clojure/proverb/1/src/proverb.clj
(ns proverb)

(defn for-want-of-a-thing-the-thing-was-lost [[thing1 thing2]]
  (list "For want of a " thing1 " the " thing2 " was lost.\n"))

(defn and-all-for-the-want-of-a-thing [thing]
  (if (nil? thing)
    ""
    (str "And all for the want of a " thing ".")))

(first '())

(defn recite [[culprit & _ :as inputs]]
  (str (->> (partition 2 1 inputs)
            (mapcat for-want-of-a-thing-the-thing-was-lost)
            (apply str))
       (and-all-for-the-want-of-a-thing culprit)))

File: solutions/java/proverb/1/src/main/java/Proverb.java
class Proverb {
    private String words[];
    private static final String FORMAT_MOST = "For want of a %s the %s was lost.\n";
    private static final String FORMAT_LAST = "And all for the want of a %s.";
    
    Proverb(String[] words) {
        this.words = words;
    }

    String recite() {
        StringBuilder out = new StringBuilder();
        
        for (int i = 0; i <= this.words.length - 2; i++) {
            out.append(String.format(FORMAT_MOST,
                                     words[i],
                                     words[i + 1]));
        }

        if (this.words.length > 0) {
            out.append(String.format(FORMAT_LAST, words[0]));
        }
        
        return out.toString();
    }

}